flang/lib/Lower/BoxAnalyzer.h

```diff
@@ -485,6 +483,19 @@ class BoxAnalyzer : public fir::details::matcher<BoxAnalyzer> {
         sym.GetType()->characterTypeSpec().length();
     if (Fortran::semantics::MaybeIntExpr expr = lenParam.GetExplicit())
       return {Fortran::evaluate::AsGenericExpr(std::move(*expr))};
+    // For assumed length parameters, the length comes from the initialization
+    // expression.
+    if (sym.attrs().test(Fortran::semantics::Attr::PARAMETER))
+      if (const auto *objectDetails =
+              sym.GetUltimate()
+                  .detailsIf<Fortran::semantics::ObjectEntityDetails>())
+        if (objectDetails->init())
+          if (const auto *charExpr = std::get_if<
+                  Fortran::evaluate::Expr<Fortran::evaluate::SomeCharacter>>(
+                  &objectDetails->init()->u))
+            if (Fortran::semantics::MaybeSubscriptIntExpr expr =
+                    charExpr->LEN())
+              return {Fortran::evaluate::AsGenericExpr(std::move(*expr))};
```

Comment:
Doesn't `toEvExpr()` work here?

Reply:
I tried but it's not accessible here. Since we do not need to clone the expr here, it's easier to just use AsGenericExpr directly.